exercises/exercise-13/database.ts

```diff
@@ -1,13 +1,79 @@
+import { fs, readline } from 'mz';
+
+type FindCriteria<T> = {
+    $eq?: T
+    $gt?: T
+    $lt?: T
+    $in?: T[]
+}
+
+type RootQuery<T> = {
+    $text?: string
+    $and?: Query<T>[]
+    $or?: Query<T>[]
+}
+
+type Query<T> = {
+    [P in keyof T]?: P extends string ? FindCriteria<T[P]> : never
+} & RootQuery<T>
+
+const getObjectKeys = <T>(o: T) => {
+    return Object.keys(o) as (keyof T)[];
+}
+
+const isSatisfyQuery = <T>(value: T, q: Query<T>, fullTextSearchFieldNames: Database<T>['fullTextSearchFieldNames']): boolean => {
+    return getObjectKeys(q).every(fieldName => {
+        if (fieldName === '$and') {
+            return q.$and && q.$and.every(andQuery => isSatisfyQuery(value, andQuery, fullTextSearchFieldNames));
```

Comment:
The `q.$and &&` part can be eliminated when using a type guard in the `if` above:
https://www.typescriptlang.org/docs/handbook/advanced-types.html

Reply:
@fabb could you elaborate, what sort of type guard do you mean considering current implementation? `fieldName` is primitive, and `$and` is an optional field in RootQuery.

Reply:
yes, it wouldn't work with the current implementation, the typeguard would need to check the type of the Query, and you would need concrete types for each query type in the Query union.
i did it like this: https://github.com/fabb/typescript-exercises/blob/my_solutions/exercises/exercise-13/database.ts#L21

Reply:
Yeah, considering your implementation that totally makes sense 👍